File: libs/language/src/lib/utils/Base58.ts
export class Base58 {
  private static readonly ALPHABET =
    '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
  private static readonly BASE_MAP = new Uint8Array(256);
  private static readonly BASE_58 = BigInt(58);

  static {
    for (let i = 0; i < Base58.BASE_MAP.length; i++) {
      Base58.BASE_MAP[i] = 255;
    }
    for (let i = 0; i < Base58.ALPHABET.length; i++) {
      Base58.BASE_MAP[Base58.ALPHABET.charCodeAt(i)] = i;
    }
  }

  public static encode(input: Uint8Array): string {
    if (input.length === 0) return '';

    let zeros = 0;
    while (zeros < input.length && input[zeros] === 0) {
      zeros++;
    }

    const encoded = [];
    let value = BigInt(0);
    for (let i = zeros; i < input.length; i++) {
      value = (value << BigInt(8)) + BigInt(input[i]);
    }

    while (value > 0) {
      const mod = value % Base58.BASE_58;
      encoded.push(Base58.ALPHABET[Number(mod)]);
      value = value / Base58.BASE_58;
    }

    while (zeros-- > 0) {
      encoded.push(Base58.ALPHABET[0]);
    }

    return encoded.reverse().join('');
  }

  public static decode(input: string): Uint8Array {
    if (input.length === 0) return new Uint8Array(0);

    let zeros = 0;
    while (zeros < input.length && input[zeros] === '1') {
      zeros++;
    }

    let value = BigInt(0);
    for (const char of input) {
      const index = Base58.BASE_MAP[char.charCodeAt(0)];
      if (index === 255) {
        throw new Error(`Invalid character found: ${char}`);
      }
      value = value * Base58.BASE_58 + BigInt(index);
    }

    const bytes = [];
    while (value > 0) {
      bytes.push(Number(value & BigInt(0xff)));
      value = value >> BigInt(8);
    }

    while (zeros-- > 0) {
      bytes.push(0);
    }

    return new Uint8Array(bytes.reverse());
  }
}

File: references/java-sdk/src/main/java/blue/language/utils/Base58.java
package blue.language.utils;

import java.math.BigInteger;

public class Base58 {
    private static final char[] ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz".toCharArray();
    private static final BigInteger BASE_58 = BigInteger.valueOf(58);

    public static String encode(byte[] input) {
        BigInteger value = new BigInteger(1, input);
        StringBuilder base58 = new StringBuilder();

        while (value.compareTo(BigInteger.ZERO) > 0) {
            BigInteger[] divmod = value.divideAndRemainder(BASE_58);
            base58.insert(0, ALPHABET[divmod[1].intValue()]);
            value = divmod[0];
        }

        // Encode leading zeros as '1's.
        int i = 0;
        while (i < input.length && input[i] == 0) {
            base58.insert(0, ALPHABET[0]);
            i++;
        }

        return base58.toString();
    }

    public static byte[] decode(String input) {
        BigInteger num = BigInteger.ZERO;
        for (char t : input.toCharArray()) {
            int p = new String(ALPHABET).indexOf(t);
            if (p == -1) {
                throw new IllegalArgumentException("Invalid character found: " + t);
            }
            num = num.multiply(BASE_58).add(BigInteger.valueOf(p));
        }

        byte[] bytes = num.toByteArray();
        boolean stripSignByte = bytes.length > 1 && bytes[0] == 0 && bytes[1] < 0;
        int leadingZeros = 0;
        for (int i = 0; i < input.length() && input.charAt(i) == ALPHABET[0]; i++) {
            leadingZeros++;
        }

        byte[] tmp = new byte[bytes.length - (stripSignByte ? 1 : 0) + leadingZeros];
        System.arraycopy(bytes, stripSignByte ? 1 : 0, tmp, leadingZeros, tmp.length - leadingZeros);
        return tmp;
    }
}